### PR TITLE
Complexity: fold map into the sum.

### PR DESCRIPTION
**Pull request by diesalbla**
Merges the use of a `map` function into the `sum` function, so as to not allocate a list of objects.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -473,12 +473,17 @@ trait Implicits {
         val syms = for (t <- tp; if t.typeSymbol.isTypeParameter) yield t.typeSymbol
         deriveTypeWithWildcards(syms.distinct)(tp)
       }
+      def sumComplexity(ts: List[Type]): Int = {
+        var s = 0
+        ts foreach { t => s += complexity(t) }
+        s
+      }
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:479]: You could make it more loopy:
```
@annotation.tailrec def g(sum: Int, xs: List[Type]): Int = xs match {
  case h :: t => g(sum + complexity(h), t)
  case _: Nil.type => sum 
}
```
Your `s` is captured, so it's off-stack. You could also consider a while loop and allocating an iterator, but that would not be as fancy.

**Review by som-snytt — CHANGES_REQUESTED** [review]
I'm not feeling well yet after a day of massive fever, so forgive me if I'm loopy. (pun)

**Review by som-snytt — APPROVED** [review]

**Comment by som-snytt** [comment]
Could also just `foldLeft` for compact expression.

**Comment by SethTisue** [comment]
Travis-CI failure is spurious (scala/scala-dev#617)

**Pull request merged**